File: LeetCode/671_Second_Minimum_Node_In_a_Binary_Tree/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    int minimumValue;
    long answer = Long.MAX_VALUE;

    public void findSecondMinimumValueHelper(TreeNode root) {
        if (root == null)
            return;

        if (minimumValue < root.val && root.val < answer) {
            answer = root.val;
        } else {
            findSecondMinimumValueHelper(root.left);
            findSecondMinimumValueHelper(root.right);
        }
    }

    public int findSecondMinimumValue(TreeNode root) {
        minimumValue = root.val;
        findSecondMinimumValueHelper(root);
        return answer == Long.MAX_VALUE ? -1 : (int) answer;
    }
}

File: LeetCode/671_Second_Minimum_Node_In_a_Binary_Tree/Solution.py
import heapq
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findSecondMinimumValue(self, root: TreeNode | None) -> int:
        queue = deque()
        visited = set()
        heap = []
        queue.append(root)

        while queue:
            node = queue.popleft()
            if node.val not in visited:
                visited.add(node.val)
                heapq.heappush(heap, node.val)

            if node.left:
                queue.append(node.left)

            if node.right:
                queue.append(node.right)

        if len(heap) < 2:
            return -1

        print(heap)
        heapq.heappop(heap)
        return heapq.heappop(heap)
